### PR TITLE
Auto mode

### PR DESCRIPTION
**Pull request by asaba96**
Works on robot. Very robust, in that it remembers the number of runs and state (error/no error).

I would like to add speed control (speed goes up/down depending on state/number runs)

In addition, Aaron had an idea of if we don't know best path, it keeps searching instead of going back to (0,0) so we don't waste a run on another search.

### CODE REVIEW

**Comment by asaba96** [comment]
In addition, as it is currently written, auto mode is on by default. We also need to set the target locations for APAC in conf.h (currently I have it set to (4,4)) and find a robust speed we want to set search and kaos at.

File: src/conf.h [hunk 1]
```diff
@@ -244,4 +251,11 @@
 
 #define SEARCH_VELOCITY 0.3
 
+//************UPDATE THESE VALUES ****************//
+#define APAC_TARGET_X 4
+#define APAC_TARGET_Y 4
```
Comment (ghost) [inline src/conf.h]: The competition maze is 16x16, and the finish is the middle 4 squares. So (7, 7) or (8, 8) would work. I think in the past we did (7, 7).

Comment (ghost) [inline src/conf.h]: Here and in other places, the spelling is APEC, not APAC.

File: src/conf.h [hunk 2]
```diff
@@ -141,6 +141,13 @@
 #define EEPROM_TARGET_X_LOCATION 516
 #define EEPROM_TARGET_Y_LOCATION 517
 
+/**********NEED LOCATIONS FOR THESE *************/
+#define EEPROM_NUM_RUNS_LOCATION 520
+#define EEPROM_STATE_LOCATION 522
```
Comment (ghost) [inline src/conf.h]: Are these safe locations? Why the message?

File: src/main.cpp [hunk 3]
```diff
@@ -74,21 +79,172 @@ void micromouse_main()
...
   Menu menu(items, false);
 
   for (;;) menu.run();
 }
 
+void autoMode()
+{
+	//MUST CLEAR MANUALLY AT VERY BEGINNING
+	PersistantStorage::setTargetXLocation(APAC_TARGET_X);
+	PersistantStorage::setTargetYLocation(APAC_TARGET_Y);
```
Comment (ghost) [inline src/main.cpp]: I would argue that we can remove this. We can upload old code to set any target location we want, and then upload the auto mode code which won't allow us to change the location anymore. In my eyes that's a good enough solution and gives us maximum flexibility for testing.

File: src/conf.h (hunk 1 above)
Comment (ghost) [inline src/conf.h]: And see my other message about how we might not even need this setting.

**Review by ghost — CHANGES_REQUESTED** [review]
Change/add the following (also made some comments on the diff):
 - Whitespace style - Change your tabs to spaces.
 - Spelling - It's APEC, not APAC :-)
 - Pauses - Like you mentioned before, we need the one-second pause in the start square to comply with the rules.

 - Question - What is maze clear? What advantage does that have over regular clear?

I haven't read through the logic of `autoMode()`, `autoRun()`, and `autoKaos()`, but I did test on the bot myself and it works pretty great. We do crash, but that's because motion sucks :-)

When we do crash, I tested turning the bot off, moving it back to the start, turning it back on, and resuming auto mode. And it works great. Nice job!!

**Comment by asaba96** [comment]
So the maze clear function clears the maze but doesn't reset numRuns and state. This is key if we see our kaos fails a lot. Often doing a maze clear and doing another run fixes the issues with Kaos crashing constantly.

**Comment by asaba96** [comment]
I will fix the issues mentioned and commit those fixes. In addition, I forgot to take out comments I left for myself. Apologies for bad styling.

File: src/conf.h (hunk 2 above)
Comment (asaba96) [reply src/conf.h]: These are safe locations. The message was for myself. I will remove it.

**Comment by ghost** [comment]
Took executive action. Rebased, added to, and merged separately.

**Pull request closed without merging**